lazyflow/request/threadPool.py

```diff
@@ -39,13 +40,32 @@ class ThreadPool:
     def __init__(self, num_workers: int):
         """Start all workers."""
         self.unassigned_tasks = queue.PriorityQueue()
+        self.ready_workers = queue.Queue()
 
-        self.workers = {_Worker(self, i) for i in range(num_workers)}
+        self.workers = [_Worker(self, i) for i in range(num_workers)]
         for w in self.workers:
             w.start()
 
         atexit.register(self.stop)
 
+        self._scheduler = threading.Thread(target=self._distribute_work)
+        self._scheduler.daemon = True
+        self._scheduler.start()
+
+    def _distribute_work(self):
+        while True:
+            try:
+                task = self.unassigned_tasks.get()
```

Comment:
Shouldn't we stop the scheduler as well?

```python
if task is STOP:
    return

....

if worker is STOP:
    return

....

def stop(...):
    ...
    self.unassigned_tasks.put_nowait(STOP)
    self.ready_workers.put_nowait(STOP)
```

Reply:
It's a daemon thread and it doesn't do any actual work, so it will gracefully shut down by itself with thread pool termination. Otherwise, we would need to add timeouts to both queues gets and introduce polling to properly terminate scheduler.

Reply:
Wouldn't putting `STOP` in both queues solve the problem?

Keeping a running thread after `ThreadPool.stop()` doesn't seem like a good behavior (it consumes resources and prevents GC for some objects), but this is probably not a big deal since we have 1 pool for the entire lifetime of a process anyway.

Reply:
And then we need to make STOP object comparable with objects in unassigned_tasks.

Reply:
It should be doable:

```python
@functools.total_ordering
class Stop:
    __lt__ = lambda _self, _other: True
    __eq__ = lambda self, other: isinstance(self, other.__class__)


STOP = Stop()
```

And we also need to fix `Request`, which should be done anyway because right now it's `__lt__` method is broken (works only with other request object):

```python
@functools.total_ordering
class Request:
    ...

    def __lt__(self, other):
        if not hasattr(other, "_priority"):
            return NotImplemented
        return self._priority < other._priority

    def __eq__(self, other):
        if not hasattr(other, "_priority"):
            return NotImplemented
        return self._priority == other._priority            
```

This is probably even better because `STOP` will appear immediately as the next element in all queues, which is what we want.

N.B.: I believe the comment about `None` in the current `Request.__lt__` is wrong: queues are populated only through `Request._wake_up` that always passes self as an argument.

Reply:
See updated revision

Reply:
Great! Explicitly passing the priority is an even better approach.